A simple, per-file metadata CRUD API

trishankatdatadog opened the pull request:
**Fixes issue #**:

#1048 

**Description of the changes being introduced by the pull request**:

A simple, per-file metadata CRUD API that should make it easy to do the most common operations, without resorting to reading and writing the entire repository all at once.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Cc @joshuagl @sechkova @MVrachev @lukpueh @woodruffw

trishankatdatadog committed:
WIP

trishankatdatadog commented:
There is so much incomplete with this draft PR, but I hope it gives the right idea. Please let me know if you vehemently disagree with the approach here!

Cc @joshuagl @sechkova and @MVrachev who have agreed to help write and polish the code

joshuagl commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,169 @@
+    # this is the tuf filename convention at the time of writing.
+    return f'{private_keypath}.pub'
+
+def get_private_keys_from_keyring(keyring: Keyring) -> KeyDict:
```
This might be cleaner as a class method on `Keyring`, i.e. `get_private_keys()`?

joshuagl commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,169 @@
+
+    return privkeys
+
+def get_public_keys_from_keyring(keyring: Keyring) -> KeyDict:
```
Similarly here, would it make more sense as a class method on `Keyring` – `get_public_keys()`?

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,148 @@
+        signed = signable['signed']
+
+        self.expiration = datetime.strptime(signed['expiration'], '%b %d %Y %I:%M%p')
+        self.version = signed['version']
```
TODO: should be able to set appropriate values for `self.consistent_snapshot` at load, also.

joshuagl reviewed (commented):
Thanks for your initial work on this @trishankatdatadog

Observation: `typing` support requires Python 3.5+ or the [typing package](https://pypi.org/project/typing/) and f-string support requires Python 3.6. 
This needn't block the initial implementation, but assuming we continue to flesh out this low-level API and use it in the refactor it might be exposed to the client, where we want to continue to support Python 2.7 and thus we can't use f-strings and may not want to add an additional dependency on typing.

So far as I can see from a preliminary review the to-do list is:
* keys
  * ~Support non ed25519 keys. Perhaps by making key type a part of: `get_private_keypath()`, `write_keypair()`, `read_keypair()`, etc?~ https://github.com/theupdateframework/tuf/pull/1060/commits/f2861bfd417df30f814cfa25128fdf5e5c2f05f8
* metadata
  * Figure out how to handle consistent snapshots
  * ~Figure out how to handle writing consistent targets~ not relevant at this low-level https://github.com/theupdateframework/tuf/pull/1060#discussion_r448908497
  * ~Add schema checks to `read_from_json()`~ https://github.com/theupdateframework/tuf/pull/1060/commits/1fbff557adec35de1fc80894cca1b2ea982c1960
  * ~Implement `update_signatures()`~ https://github.com/theupdateframework/tuf/pull/1060/commits/5ef60ca1871953d3982458acda736815a4c9bc86
  * Implement `signable()` for each of:
    * ~`Timestamp`~ https://github.com/theupdateframework/tuf/pull/1060/commits/11d76e72ddbd1dafab70c35213db30396dcb7e44
    * ~`Snapshot`~ https://github.com/theupdateframework/tuf/pull/1060/commits/92f97a45e3c8934b2872075fe53e606f459aad22
    * ~`Targets`~ https://github.com/theupdateframework/tuf/pull/1060/commits/916055aa54452283831db0fc0d351e368c8b4ca0
  * Implement `update()` for each of 
    * ~`Timestamp`~ https://github.com/theupdateframework/tuf/pull/1060/commits/11d76e72ddbd1dafab70c35213db30396dcb7e44
    * ~`Snapshot`~ https://github.com/theupdateframework/tuf/pull/1060/commits/92f97a45e3c8934b2872075fe53e606f459aad22
    * ~`Targets`~ https://github.com/theupdateframework/tuf/pull/1060/commits/db0f8a73a70e2dd7103769977a9f939a500f1f94
  * ~Add method to `Metadata` to bump expiration by a delta, `Metadata.bump_expiration(delta='1d')`~
    * ~It would be easy for the implementation if this took a [timedelta](https://docs.python.org/3/library/datetime.html#datetime.timedelta) object, i.e. `Metadata.bump_expiration(delta=timedelta(days=1))` – I'd appreciate some thoughts on how useful/usable such an API is?~ https://github.com/theupdateframework/tuf/pull/1060/commits/721def4eb06acf6e7c2accf2d7201541f0819d71
  * ~Add method to `Metadata` to bump version number; `Metadata.bump_version()`~ https://github.com/theupdateframework/tuf/pull/1060/commits/721def4eb06acf6e7c2accf2d7201541f0819d71
* ~Use `StorageBackendInterface`~ https://github.com/theupdateframework/tuf/pull/1060/commits/0ca471ed2ad151c61647400e3ac07670dc621b9e
* ~We should raise sslib.FormatError or ValueError instead of AssertionErrors (for input validation)~ https://github.com/theupdateframework/tuf/pull/1060/commits/76cb560a4601dc264390d10143a68cc05d16787f
* Add docstrings
* Add unit tests – WIP here https://github.com/theupdateframework/tuf/pull/1060/commits/57c98d45ac570b6947420b9f535a3765c610c0c5 and here https://github.com/theupdateframework/tuf/pull/1060/commits/34a0680947ab5e0afc85572e438b40bcb98cd98a
* ~Streamline `metadata` interaction model so that `signatures` and `signed` are only ever accessed as a property, like `signable`~
* ~make as many thing as possible private variables of the objects~ https://github.com/theupdateframework/tuf/pull/1060/commits/76cb560a4601dc264390d10143a68cc05d16787f

trishankatdatadog commented:
Hi Joshua,

> Observation: `typing` support requires Python 3.5+ or the [typing package](https://pypi.org/project/typing/) and f-string support requires Python 3.6.
> This needn't block the initial implementation, but assuming we continue to flesh out this low-level API and use it in the refactor it might be exposed to the client, where we want to continue to support Python 2.7 and thus we can't use f-strings and may not want to add an additional dependency on typing.

Agreed. I left the typing in there as to help communicate intent between different developers 🙂 So you can think of it as scaffolding we can remove before hitting production...

> Add bump_role() API somewhere?

Aha, it's not clear, sorry, but it's supposed to be a part of [`Metadata.sign()`](https://github.com/theupdateframework/tuf/pull/1060/files#diff-d2aa622cc58ad47011c52d0adc99323dR73). Let me know whether this is a good idea or not...

How should we divide-and-conquer going forward?

woodruffw commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,169 @@
+
+class Threshold:
+
+    def __init__(self, m: int = 1, n: int = 1):
```
Tiny nit, but can we use more descriptive parameter names + attributes here? I know it's an m-of-n scheme but `minimum` and `keyholders` (or similar) would be a little easier to read 🙂

lukpueh commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,169 @@
+        self.m = m
+        self.n = n
+
+class Keypath:
```
Nit: `KeyPath`

lukpueh commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,169 @@
+        self.path = path
+        self.obj = obj
+
+class Keypair:
```
Nit: `KeyPair`

lukpueh commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,169 @@
+
+Keypairs = List[Keypair]
+
+class Keyring:
```
Nit: `KeyRing`

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,148 @@
+
+    # TODO: We need to update the expiration timestamp using self.expiration.
+    # Oh, and bump the version number.
+    # And, oh, take care of consistent snapshot of metadata.
```
I think none of this should be done in the sign function. Let's have a separate API function (or set of functions) for this.

joshuagl replied:
@lukpueh beat me to it. I wrote up my though process around this a bit in a comment below.

lukpueh commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,169 @@
+
+    return Keypair(private_key, public_key)
+
+def rename_keys_to_match_keyid(keystore_dir: str, keypair: Keypair) -> None:
```
This seems like a crutch that we only need because we don't write the keys to the right location in the first place. Maybe we can rather fix this at the source of that problem (in sslib) instead of working around here?

lukpueh reviewed (commented):
Great work, @trishankatdatadog! I especially like the scaffolding in `metadata.py`.

A left a few inline comments and here are two more high-level observations:
- We should raise sslib.FormatError or ValueError instead of AssertionErrors (for input validation)
- When we add docstrings, let's please use a standard format
(see last item of "Issues with the current guideline" in https://github.com/secure-systems-lab/code-style-guidelines/issues/20 and https://github.com/in-toto/in-toto/pull/369)

joshuagl commented:
> > Observation: `typing` support requires Python 3.5+ or the [typing package](https://pypi.org/project/typing/) and f-string support requires Python 3.6.
> > This needn't block the initial implementation, but assuming we continue to flesh out this low-level API and use it in the refactor it might be exposed to the client, where we want to continue to support Python 2.7 and thus we can't use f-strings and may not want to add an additional dependency on typing.
> 
> Agreed. I left the typing in there as to help communicate intent between different developers 🙂 So you can think of it as scaffolding we can remove before hitting production...

I assumed it was something like that, but thought it worth explicitly noting.

> > Add bump_role() API somewhere?
> 
> Aha, it's not clear, sorry, but it's supposed to be a part of [`Metadata.sign()`](https://github.com/theupdateframework/tuf/pull/1060/files#diff-d2aa622cc58ad47011c52d0adc99323dR73). Let me know whether this is a good idea or not...

Ah, understood! We'll need a way to set/pass a delta by which to increase the expiration, which doesn't feel like it fits naturally into the method signature of `sign()` and is even less clean in the method signature of `write_to_json()`, from which `sign()` is called.

I thought a delta by which to bump the `expiration` could become a property of the `Metadata` object (i.e. PEP 458 [recommends specific expiry deltas per role](https://www.python.org/dev/peps/pep-0458/#metadata-expiry-times) and this would map nicely) however the delta isn't a part of the metadata itself and therefore couldn't be loaded into the object by reading the on-disk representation... therefore – this wouldn't work.

As this is defined as a low-level (or granular?) API, perhaps we should just be explicit and have functions for incrementing version number and increasing expiration. We can build convenience API around these in future as we flesh this out?

joshuagl committed:
tuf.api: set consistent_snapshot during read_from_json

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: add helpers to bump version and expiration

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: implement update_signatures()

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: implement metadata.Snapshot

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
> How should we divide-and-conquer going forward?

The to-do list I [sketched above](https://github.com/theupdateframework/tuf/pull/1060#pullrequestreview-439026780) has many of the items broken down by module, perhaps you could tackle `keys` while we work on `metadata`?

Then we can hammer out typing, docstrings, tests, coding style and assertions as a follow-on task?

joshuagl committed:
tuf.api: fix missing ':' in keys

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: fix imports in metadata

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: fix version check in metadata

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: fix loading expiration

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
WIP tests for tuf.api.metadata

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
I've got some WIP changes https://github.com/joshuagl/tuf/commits/joshuagl/simple-update-api%2BMVrachev-length-hashes-optional and have updated the [to do list](https://github.com/theupdateframework/tuf/pull/1060#pullrequestreview-439026780) to indicate what I have done. 

Note: I'm using a helper from #1031 so my branch includes a merge of that PR

joshuagl committed:
tuf.api: WIP implement Timestamp

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
all of the code I've written so far assumes that the sequence of events is:
```Python
metadata = metadata.Metadata()
metadata.read_from_json('/path/to/metadata.json')
```
and therefore `self.signatures` and `self.signed` exist and are meaningful. That means we can't currently instantiate an object, populate it with code and write it to disk. We can only load and modify. That seems undesirable?

Should we smooth out the interaction model so that `signatures` and `signed` are only ever accessed as a property, like signable? 


Additionally, should we make `read_from_json()` a class method that can be used to create an instance of an object by loading a file?
```Python
snapshot = Snapshot.read_from_json('/path/to/snapshot.json')
```

trishankatdatadog commented:
> Should we smooth out the interaction model so that `signatures` and `signed` are only ever accessed as a property, like signable?

Yes, we should. Also, `signatures` and `signed` should be objects under `signable`, no?
 
> Additionally, should we make `read_from_json()` a class method that can be used to create an instance of an object by loading a file?

💯

trishankatdatadog committed:
much simpler keys

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

sechkova committed:
tuf.api: implement metadata.Targtes.read_from_json()

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
tuf.api: implement metadata.Targets.update()

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
tux.api: implement metadata.Targets.signable()

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Skip an optional keyword in the schema

Skip a keyword if it is optional in the schema and the value
passed in is set to None.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
tuf.api: simplify metadata.Targets.signable()

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented on tuf/api/metadata.py:
From the initial API design I assume that it covers only the case of using existing fileinfo. If this is true then writing consistent targets is not applicable here:
https://github.com/theupdateframework/tuf/blob/de8649e12a850fccbc745f0129d43956ef7da33b/tuf/repository_lib.py#L155

If we want to add the option to NOT use existing fileinfo and using tuf for calculating hashes, we need to extend the api (maybe on a later pass) with targets paths and storage_backend.

sechkova committed:
tuf.api: add basic schema checks in read_from_json

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
tuf.api: use StorageBackendInterface

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

joshuagl committed:
tuf.api: make expires a property

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: make signed & signatures properties of Metadata

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: instantiate objects from the JSON

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
More tests tuf.api and verify data!

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: convert keys.Algorithm to dict

Enum isn't available in Python 2.7 and accessing the enum members
as items, i.e. Algorithm['RSA'] throws KeyError when the member
is a method.
Work around both of these issues by converting to a dict.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: shuffle up signed/signatures/signable interface

* Add signed_bytes property to return a canonicalised version of signed
* Make the signed property responsible for converting the object into the
  dict format expected of the metadata, requiring a per-class
  implementation
* Make signable a generic base-class property which returns a dict of
  signatures and signed

With these changes we can load and verify metadata with the low-level API!

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
> > Should we smooth out the interaction model so that `signatures` and `signed` are only ever accessed as a property, like signable?
> 
> Yes, we should. Also, `signatures` and `signed` should be objects under `signable`, no?

I haven't implemented this suggestion yet as I was worried about the complexity increase of nested classes. I'd be curious to hear/read what everyone thinks of the current implementation.

trishankatdatadog commented:
> I haven't implemented this suggestion yet as I was worried about the complexity increase of nested classes. I'd be curious to hear/read what everyone thinks of the current implementation.

This is simply looking fabulous, thanks very much, Joshua! 💯 

I have a minor quibble about how we should make as many thing as possible private variables of the objects, but I started that, so it's on me, and it's also easy to fix 🙂

sechkova committed:
WIP added tuf.api keys tests

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

joshuagl commented:
> > I haven't implemented this suggestion yet as I was worried about the complexity increase of nested classes. I'd be curious to hear/read what everyone thinks of the current implementation.
> 
> This is simply looking fabulous, thanks very much, Joshua! 💯

You're very kind, thanks!

> I have a minor quibble about how we should make as many thing as possible private variables of the objects, but I started that, so it's on me, and it's also easy to fix 🙂

I completely agree. I've added it to the list of things to address in https://github.com/theupdateframework/tuf/pull/1060#pullrequestreview-439026780

joshuagl commented on tests/test_tuf_api.py:
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
+
+# Copyright 2014 - 2017, New York University and the TUF contributors
```
Clearly the wrong date range. 🤦

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,236 @@
+
+    @property
+    def expires(self) -> str:
+        return self.expiration.replace(tzinfo=None).isoformat()+'Z'
```
In https://github.com/theupdateframework/tuf/pull/1012#discussion_r403039950 @woodruffw wonders whether `unix_timestamp_to_datetime` should just include the Z suffix by default if we always intend times to be GMT.

trishankatdatadog commented:
@joshuagl I think we're largely done, no? Should we get @woodruffw to start alpha-testing? 🧪

joshuagl committed:
tuf.api: treat all datetime's as UTC

We don't capture timezone information in the metadata, therefore we should
not capture it in the interfaces. Ensure we remove timezone information
from any datetime objects when they are assigned to the expiration
property of a Metadata object.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
tuf.api: drop use of dateutil

All of the functionality we need is available from the standard library
which reduces our dependency footprint. Having minimal dependencies is
especially important for update clients which often have to vendor their
dependencies.

However, dateutil.relativedelta is richer than timedelta and helps to
provide a clearer API. For example, with relativedelta it's possible
to specify a delta in years *and* dateutil will do the right thing for
leap years.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Test Metadata.bump_expiration() with relativedelta

dateutil provides an interface which is much easier to reason about for
users, i.e. it provides an interface for year deltas which automatically
handles leap years. Add some tests to try and ensure that, even though it
uses standard library functionality, the metadata API can accept
dateutil.relativedelta and do the right thing.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
> @joshuagl I think we're largely done, no? Should we get @woodruffw to start alpha-testing? 🧪

I'd certainly appreciate some user feedback on this API from @woodruffw !

The to do list before we can remove the draft flag from this PR is:
* Figure out if/how to handle consistent snapshots
  * maybe this would be at the `write_to_json()` level? If `consistent_snapshot` is true we could massage the passed filename into the expected format for a consistent snapshot? Is that too magical? Should consistent snapshot handling be done at a higher level?
* raise sslib.FormatError or ValueError instead of AssertionErrors (for input validation)
* Add docstrings
* Finish unit tests, ensure good (90%+?) coverage

joshuagl commented:
The interface for metadata currently has two very similar properties:
* `Metadata.expiration` can be used to set/get the `expiration` `datetime` object
* `Metadata.expires` can be used to get the string representation of the `expiration` `datetime` object that we will write to the metadata

The names are a bit too similar for my taste, some thoughts:
* should we make the `expires` property more explicitly named: `expiration_string`, or similar?
* should we make `expires` an internal helper – will users of the public API ever need the string representation?
* if so, should we add a setter for `expires` so that users can set the `expiration` `datetime` from an iso8601 string representation?

trishankatdatadog commented:
> * maybe this would be at the `write_to_json()` level? If `consistent_snapshot` is true we could massage the passed filename into the expected format for a consistent snapshot? Is that too magical? Should consistent snapshot handling be done at a higher level?

I think maybe we should pass a directory instead of filename to `write_to_json()` and if the `consistent_snapshot` flag is set, we can write an additional file in there. WDYT?

We also need a similar interface for writing consistent targets...

trishankatdatadog commented:
> * if so, should we add a setter for `expires` so that users can set the `expiration` `datetime` from an iso8601 string representation?

I like this option the best, but we should provide a more friendly interface that takes a `relativedelta` as an input instead of a `datetime`

joshuagl commented:
> > * maybe this would be at the `write_to_json()` level? If `consistent_snapshot` is true we could massage the passed filename into the expected format for a consistent snapshot? Is that too magical? Should consistent snapshot handling be done at a higher level?
> 
> I think maybe we should pass a directory instead of filename to `write_to_json()` and if the `consistent_snapshot` flag is set, we can write an additional file in there. WDYT?

That seems like a sound proposal to me.

> We also need a similar interface for writing consistent targets...

Should we be handing consistent targets at this low-level API? At least in Warehouse we will be using existing fileinfo and not interacting with the targets directly, should we assume that a higher-level interface does the right thing with consistent targets?

joshuagl commented:
> > * if so, should we add a setter for `expires` so that users can set the `expiration` `datetime` from an iso8601 string representation?
> 
> I like this option the best, but we should provide a more friendly interface that takes a `relativedelta` as an input instead of a `datetime`

Just to be clear, are you advocating for public properties for setting/getting expiration as a `datetime` and an iso8602 string?

I made some assumptions that this low-level API might eventually end up as part of the client code and so explicitly opted to drop `dateutil` in favour of the standard library (https://github.com/theupdateframework/tuf/pull/1060/commits/54e1f9c03b09bdf05fd0a3cd3e4db35c6480c5a7) to reduce the number of dependencies that i.e. Pip will have to vendor (see #1065).
However, I did add a test and `dateutil.relativedelta` seems to behave as expected, see https://github.com/theupdateframework/tuf/pull/1060/commits/42372872d2a7f83bc4abad34d3f52a1d606fd786.

trishankatdatadog committed:
minor edits

* classmethod for init RAMKey from file
* private class variables
* more typing for methods
* better names for arguments

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

lukpueh commented:
> - raise sslib.FormatError or ValueError instead of AssertionErrors (for input validation)

I lean towards consistently using [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) and [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError) for bad arguments, since this is what they seem to be meant for.

The only advantage I can see of using a custom sslib's `FormatError` is that our own or user code can easily distinguish our Value/TypeErrors from those of other code. But I don't think that is necessary. Ideally, we validate arguments at the user boundary, so the user should know that we raised the errors, and all errors from further down the stack are our responsibility anyway, even if we didn't raise them. What do others think?

On a very related side-note, if we move away from our custom `FormatError` we also need to re-consider our usage of `check_match`, which anyways would be a good idea for the various reasons outlined in https://github.com/secure-systems-lab/securesystemslib/issues/183. :)

lukpueh commented:
> - Add docstrings

I volunteer.

lukpueh commented:
Regarding expiration:

IMO the class should have an `expires` attribute to align with the json metadata representation. The data type of that attribute could then either be a string in the required iso8602 format to fully align with the json metadata representation,  or a datetime object for easier further use. At any rate, we should have additional convenience methods with mnemonic names to set/get the date in the desired format.

FWIW in in-toto I keep the expires attribute as iso8602 string and provide a [`set_relative_expiration`](https://github.com/in-toto/in-toto/blob/6cfa36cfc2c4b864936560294c032d3bfa869064/in_toto/models/layout.py#L150) convenience method.

lukpueh commented:
> > Should consistent snapshot handling be done at a higher level?
> 
> I think maybe we should pass a directory instead of filename to `write_to_json()` and if the `consistent_snapshot` flag is set, we can write an additional file in there. WDYT?

I like both your ideas. Can we keep `write_to_json` as is, and add a wrapper that does what Trishank says? E.g....
```python
def write(self, metadata_dir, storage_backend=None):
    self.write_to_json(
        os.path.join(metadata_dir, self._name + ".json")

    if self._consistent_snapshot: # ... or isinstance(Root) # not yet relevant
        self.write_to_json(
                os.path.join(metadata_dir, self._version + "." + self._name  + ".json")
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,250 @@
+class Metadata(ABC):
+    # By default, a Metadata would be a rather empty one.
+    def __init__(self, consistent_snapshot: bool = True, expiration: datetime = datetime.today(), keyring: Optional[KeyRing] = None, version: int = 1) -> None:
+        self.__consistent_snapshot = consistent_snapshot
```
I am not sure if consistent_snapshot should be an attribute of the base metadata class. In TUF the consistent_snapshot attribute is only available on the root role metadata.

I'd really prefer if we only kept things that belong to the corresponding TUF role on a given metadata class. 

What do others think about an additional management class? Something like `Repository` but without implying that a full repository is required. We can also punt on the decision, where this is info comes from and just require the user of this interface to pass `consistent_snapshot` to the method that needs it (e.g. for writing metadata to disk).

trishankatdatadog replied:
> We can also punt on the decision, where this is info comes from and just require the user of this interface to pass consistent_snapshot to the method that needs it (e.g. for writing metadata to disk).

Good question. Personally, I agree with this right now, until we figure out something better long-term.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,250 @@
+    def __init__(self, consistent_snapshot: bool = True, expiration: datetime = datetime.today(), keyring: Optional[KeyRing] = None, version: int = 1) -> None:
+        self.__consistent_snapshot = consistent_snapshot
+
+        self.__keyring = keyring
```
I think we should avoid storing verification keys for a role on that role. See #660.

trishankatdatadog replied:
I think I see what you mean, especially in the context of delegations. But how else should we conveniently/naturally sign/verify metadata? It is assumed here that the user can figure out the right keys to use. Is this a bad assumption?

trishankatdatadog replied:
Oh, I think I see what you're really saying now. Yeah, sure, we can pass it in per sign/verify call, I guess, less convenient, but maybe philosophically cleaner

lukpueh commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,98 @@
+
+class KeyRing:
+
+    def __init__(self, threshold: Threshold, keys: Keys):
```
Which threshold is this?

trishankatdatadog replied:
Oh, I think I see what you mean. Again, context-dependent, and up to the user to figure out.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,250 @@
+        if not updated:
+            signatures.append(keyid_signature)
+
+    def sign(self) -> JsonDict:
```
I'd rather pass the signing key(s) to this function instead of having them stored on the object. (see #660)

trishankatdatadog replied:
See my comment above

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,250 @@
+        self.__signatures = signatures
+        return self.signable
+
+    def verify(self) -> bool:
```
I'd rather pass the verification key(s) to this function instead of having them stored on the object. (see #660)

trishankatdatadog replied:
See my comment above

lukpueh commented on tuf/api/metadata.py:
Maybe a stupid question, but is it a common pattern to first create the parent class object and then the child from it?

trishankatdatadog replied:
I guess we could make `__init__()` read a file, if provided, and use default versions when none is provided?

lukpueh commented on tuf/api/metadata.py:
Does this work? I think `targets` is not of type `Targets` here, but of type `Metadata`.

trishankatdatadog replied:
Sure, I can we can say `Targets` instead of `Metadata` here w/o breaking `typing`

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,250 @@
+    @property
+    @abstractmethod
+    def signed(self) -> JsonDict:
+        raise NotImplementedError
```
Do we really need abc/abstractmethod/NotImplementedError, etc... in our Metadata base class? IIUC the list of possible subclasses is: root, snapshot, timestamp and targets. I don't think we need to plan for users implementing arbitrary subclasses.

trishankatdatadog replied:
It was just a way to denote that this method must be overridden

lukpueh commented:
Thanks for getting back to me about my review comments, @trishankatdatadog, and generally thanks a lot for the hard work,  everyone!

trishankatdatadog commented:
> Thanks for getting back to me about my review comments, @trishankatdatadog, and generally thanks a lot for the hard work, everyone!

Thanks, and where should we go from here? 🙂 Specifically, what can I help with?

Commit:
Quick-fix programming errors in api.keys module

- Finalize min/max -> least/most refactor
- Comment out unclear input validation
- Use string literal for foward referencing type hint
(see https://www.python.org/dev/peps/pep-0484/#forward-references)

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Make Metadata a container class (WIP)

This commit performs restructuring on the recently added metadata
class model architecture, which shall be part of a new simple TUF
API.

The key change is that the Metadata class is now used as container
for inner TUF metadata (Root, Timestamp, Snapshot, Targets) instead
of serving as base class for these, that means we use 'composition'
instead of 'inheritance'. Still, in order to aggregate common
attributes of the inner Metadata (expires, version, spec_version),
we use a new baseclass 'Signed', which also corresponds to the
signed field of the outer metadata container.

Based on prior observations in TUF's sister project in-toto, this
architecture seems to more closely represent the metadata model as
it is defined in the specification (see in-toto/in-toto#98 and
in-toto/in-toto#142 for related discussions).

Note that the proposed changes require us to now access some
attributes/methods via the signed attribute of a Metadata object
and not directly on the Metadata object, but it would be possible
to add short-cuts. (see todo notes in doc header).

Further changes include:
 - Add minimal doc header with TODO notes

 - Make attributes that correspond to fields in TUF JSON metadata
public again. There doesn't seem to be a good reason to protect
them with leading underscores and use setters/getters instead, it
just adds more code.

 - Generally try to reduce code.

 - Remove keyring and consistent_snapshot attributes from metadata
   class. As discussed in #1060 they are a better fit for extra
   management code (also see #660)

- Remove sslib schema checks (see TODO notes about validation in
  doc header)

 - Drop usage of build_dict_conforming_to_schema, it seems a lot
   simpler and more explicit to just code this here.

 - ... same goes for make_metadata_fileinfo

 - Adapt tests accordingly

TODO: Document!!!
Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

joshuagl commented on tuf/api/metadata.py:
if it is `None`, do we need to delete an existing value?

lukpueh replied:
Good question. Is this something you and @MVrachev discussed on #1031? What do you think should happen? I understood passing `length=None` as, "I don't want length on this".  But it could also mean, "I don't want to re-assign length, use the old one."

joshuagl replied:
We did not. It seemed more apparent here because my imagined workflows for the API are different. This API feels more focused on loading and updating existing metadata, such that we'll often start from a populated class structure and update it. Whereas repository_tool feels like it's more about generating new metadata.

I think passing `length=None` should mean "I don't want a length on this", but we should think about the (unlikely?) case that i.e. a timestamp with a length is loaded, and the next version doesn't want a length.

lukpueh replied:
Not sure I fully understand. We do override the existing *fileinfo* in L260. So any previously loaded `version`, `length` and `hashes` will be discarded at any rate.

joshuagl replied:
🤦 if we're not updating the existing values what I wrote makes no sense.

lukpueh replied:
No worries! I thought you meant that someone wanted to update version only and keep the existing hashes and length. Right now this would require to do something like:
```python
obj.update(version=version, length=obj.length, hashes=obj.hashes)
```
But I think that's okay.

joshuagl commented on tuf/api/metadata.py:
Same, what if hashes is `None` (but wasn't before)?

joshuagl commented on tuf/api/metadata.py:
this error message doesn't match the conditional

lukpueh replied:
Good catch. Let's keep the note around and I'll fix it together with other validation updates. What do you think about the related ideas I floated in the doc header todo list?

joshuagl replied:
I'll comment on the ideas you floated in-line 😃

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -126,25 +92,27 @@ def __update_signature(self, signatures, keyid, signature):
-    def sign(self) -> JsonDict:
+    def sign(self, key_ring: KeyRing) -> JsonDict:
+        # FIXME: Needs documentation of expected behavior
         signed_bytes = self.signed_bytes
```
unused variable?

lukpueh replied:
Oh, @trishankatdatadog passed `signed_bytes` to `key.sign`. I didn't understand why (same goes for `signatures = self.__signatures`) and started changing it, but then though it might be better to ask him why he did it that way. :) @trishankatdatadog, could you help me understand the `sign` and `verify` functions?

trishankatdatadog replied:
You mean `Metadata.(sign|verify)`? To call `Key.(sign|verify)`, really, it should not be concerned with how keys actually do the signing. I have no strong opinions on where `signed` and `signatures` come from, please feel free to adjust according to your new implementation.

joshuagl commented on tuf/api/metadata.py:
```suggestion
            storage_backend: Optional[StorageBackendInterface] = None) -> None:
```

lukpueh commented:
> > * Add docstrings
> 
> I volunteer.

So I haven't yet had a chance to write the docstrings... 😬 

But following some discussions with @joshuagl I shuffled the metadata model class architecture quite a bit.  My intention was to make it even simpler and above all to make the classes match the metadata format specification more closely. An overview of the changes + rationale can be found in the commit message of 3e249f5bdd9e600475a05eab3587c5388eeedc3a.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -1,3 +1,38 @@
+ * Finalize/Document Verify/Sign functions (I am not fully sure about expected
+   behavior)
+
+ * Validation (some thoughts ...)
+   - Avoid schema, see secure-systems-lab/securesystemslib#183
+   - Provide methods to validate JSON representation (at user boundary)
+   - Fail on bad json metadata in read_from_json method
+   - Be lenient on bad/invalid metadata objects in memory, they might be
+     work in progress. E.g. it might be convenient to create empty metadata
+     and assign attributes later on.
+   - Fail on bad json metadata in write_to_json method, but with option to
+     disable check as there might be a justified reason to write WIP
+     metadata to json.
```
💯 This all sounds sensible to me.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -1,3 +1,38 @@
+     disable check as there might be a justified reason to write WIP
+     metadata to json.
+
+ * It might be nice to have short-cuts on the Metadata class to methods and
+   attributes of the contained Signed object. If we do this, we should only do
+   it on common methods/attributes (e.g. version, bump_version, expires,
+   bump_expiration)
```
I'm undecided on this. I'm leaning in favour of not mapping attributes of the Signed object into the metadata class, because something like `timestamp.signed.expires` better matches the metadata format in the specification and that's what we set out to achieve with this re-working effort.

lukpueh replied:
Yeah, I'm a bit torn too. I think as long as it is clear that the `expires`, `version`, etc. are only convenience short-cuts/aliases to those attributes defined on the inner metadata it is okay. 

Still, it might be confusing to have aliases on the common attributes only, but not on those specific to any of timestamp, snapshot, targets... and adding conditional shortcuts to specific fields too might make the short-cutting implementation to complicated.

So if we agree that the typing effort (as in amount of used characters) of e.g. `timestamp.signed.expires` is still reasonable, we can go without aliases for the time being.

Besides, it might make sense to adopt a workflow, where we only work on Metadata objects when needed, and otherwise work directly on the inner metadata, e.g.:
```python

timestamp_md = Timestamp.load_from_json("timestamp.json")
timestamp_md.verify(keys)

# Discard container once existing signatures are verified and
# do all the heavy lifting directly on timestamp
timestamp = timestamp_md.signed
# timestamp.foo()
# timestamp.bar()
# ...

# Re-wrap timestamp in Metadata when signing is needed,
timestamp_md = Metadata(signed=timestamp)
timestamp_md.sign(keys)
timestamp_md.write_to_json()
```

joshuagl replied:
> So if we agree that the typing effort (as in amount of used characters) of e.g. `timestamp.signed.expires` is still reasonable, we can go without aliases for the time being.

I vote we go without aliases to start. In terms of API stability, it's easier to add them later if they are needed than it is to remove them.

> Besides, it might make sense to adopt a workflow, where we only work on Metadata objects when needed, and otherwise work directly on the inner metadata, e.g.:
> 
> ```python
> timestamp_md = Timestamp.load_from_json("timestamp.json")
> timestamp_md.verify(keys)
> 
> # Discard container once existing signatures are verified and
> # do all the heavy lifting directly on timestamp
> timestamp = timestamp_md.signed
> # timestamp.foo()
> # timestamp.bar()
> # ...
> 
> # Re-wrap timestamp in Metadata when signing is needed,
> timestamp_md = Metadata(signed=timestamp)
> timestamp_md.sign(keys)
> timestamp_md.write_to_json()
> ```

Nice idea. We should certainly give ourselves time to experiment with this pattern and see how it feels in use.

lukpueh replied:
> I vote we go without aliases to start. In terms of API stability, it's easier to add them later if they are needed than it is to remove them.

Agreed.

> Nice idea. We should certainly give ourselves time to experiment with this pattern and see how it feels in use.

Forgot to mention, this is what we e.g. do in [layout in-toto's verification routine](https://github.com/in-toto/in-toto/blob/a621c3bf3cbe1dd442b2388c685b3130f032430b/in_toto/verifylib.py#L1462-L1465).

trishankatdatadog replied:
I like this idea. Feels a lot better than trying to update signatures in place. Most importantly, we should ask what @woodruffw thinks...

woodruffw replied:
Sorry for the late response -- I agree with @joshuagl here.

trishankatdatadog commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,98 @@
+        self.__obj = obj
+
+    @classmethod
+    def read_from_file(cls, filename: str,  algorithm: str, passphrase: Optional[str] = None, storage_backend: Optional[StorageBackendInterface] = None) -> Key:
```
We don't need `storage_backend` here (my bad).

trishankatdatadog committed:
fix badly translated assertions

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
Not all Keys read from files

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

lukpueh commented:
Here is a summary of my thoughts about signing, signature verification and key handling in regards to this API.

**Keys**
Neither signing keys nor verification keys should be attributes of the role, i.e. `Metadata.signed` instance, they are meant to sign or verify. Instead and in order to align with the role model of the specification, they should be attributes of the *delegating* (this includes root) role. Same goes for the signature threshold. 
Here are two suggestions how to design the signing (or verification) interface.

~~delegating_role.sign(delegated_role) # delegating role signs the delegated role~~ (see https://github.com/theupdateframework/tuf/pull/1060#issuecomment-660108655)
```python
delegated_role.sign(keys) # delegated role signs itself with keys, e.g. from delegating_role, or a detached keyring
```
Caveat: 
- Although signing keys conceptually belong to the delegating role (as in *"delegating role signs for delegated role*"), they must never be part of the corresponding role metadata.
- This API might be used to sign delegated role metadata without having (or requiring) full knowledge of the delegating role (only the delegating role's keys are needed).


**Sign**
This PR currently proposes a sophisticated way of updating signatures by keyids. Although this seems convenient, I am not sure if there is actually a use case. In order to keep the interface minimal, I suggest to implement a `sign` function that just appends the signature, and leave it up to the caller if they want to first remove all or a few signatures from the list.

For comparison, [here is how we do this in in-toto](https://github.com/in-toto/in-toto/blob/a621c3bf3cbe1dd442b2388c685b3130f032430b/in_toto/models/metadata.py#L137-L163).

**Verify** 
I think it would be good two provide two functions, one that only verifies if the signature(s) on the object are valid, and another one that checks if enough validly verified signature are available to meet a threshold.

In-toto has the former as [method on the metadata class](https://github.com/in-toto/in-toto/blob/a621c3bf3cbe1dd442b2388c685b3130f032430b/in_toto/models/metadata.py#L198-L256) and the latter as [part of the full in-toto final product verification routine](https://github.com/in-toto/in-toto/blob/a621c3bf3cbe1dd442b2388c685b3130f032430b/in_toto/verifylib.py#L376-L387).

lukpueh commented:
> - Although signing keys conceptually belong to the delegating role (as in "delegating role signs for delegated role"), they must never be part of the corresponding role metadata.

I'm actually not so sure about this. A signing key probably does not need to belong to the delegating role, it could come from anywhere. So I suggest we have the caller decide where to get it from, and just pass it to the signing function as argument.

And we should provide the same workflow for the simple signature verification function, i.e. passing an unaffiliated key to just verify a signature. Even if, in the course of the client verification, the key (and threshold) are taken from the delegating role.

trishankatdatadog commented:
> Here is a summary of my thoughts about signing, signature verification and key handling in regards to this API.

Agree with your thoughts. We could write some convenience functions later to make it easier to find and associate the right Keys with the right Metadata, but we don't need them for now.

joshuagl commented:
I was just writing a similar comment to @trishankatdatadog's. Really enjoying seeing this API evolve as a collaborative effort, thanks all!

trishankatdatadog commented:
> I was just writing a similar comment to @trishankatdatadog's. Really enjoying seeing this API evolve as a collaborative effort, thanks all!

Likewise 🙂 Hoping to push soon that new type of Key I talked about yesterday... 🤞

trishankatdatadog committed:
sign/verify w/ Vault

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog commented:
Good news everyone: with 9285727ec50d1bb746c980525efc03d647f9925f, we finally have a Vault-TUF integration, where metadata can be signed and verified using ECDSA/Ed25519/RSA keys kept entirely in Hashicorp Vault 🚀 

It is a decent start, but there are a few important TODOs left, notably:
- [x] Test that py-TUF can parse public keys produced by this integration
- [x] Test that py-TUF can verify signature produced by this integration
- [ ] Start Vault, get Vault address and root token, and enable Transit Secrets Engine in unit tests
- [ ] Allow passing in a prehashed input instead of the whole input

trishankatdatadog committed:
fix oversight in test

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
update and automate requirements

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
add P-384 and RSA-3072

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
Python verification of signatures from Go

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
minor comments

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog commented:
Before we finalize, I also think we should consider the notion of [unbounded vs bounded keys](https://github.com/in-toto/in-toto-golang/pull/56#issuecomment-662689093).

joshuagl commented:
The Vault integration sounds very interesting, great work @trishankatdatadog. 💯 

I wonder whether a key management interface would be better suited to being integrated at the securesystemslib level, rather than TUF?

Even if TUF is the more appropriate home, I think it would be better to keep this PR focused on introducing the CRUD API. With any extra functionality built on top of that introduced in future PR's. What do you think?

On the CRUD API; @woodruffw did you have chance to review the proposed API and see how well it suits your needs for Warehouse integration?

trishankatdatadog commented:
> Even if TUF is the more appropriate home, I think it would be better to keep this PR focused on introducing the CRUD API. With any extra functionality built on top of that introduced in future PR's. What do you think?

No problemo. I will let @lukpueh, @woodruffw, and yourself decide.

woodruffw commented on tuf/api/keys.py:
```diff
@@ -0,0 +1,427 @@
+
+class VaultKey(Key):
+
+    class AuthenticationError(Exception): pass
```
Tiny nit, but perhaps don't nest this exception class or the following enums?

Having the former in a standalone `exceptions` module and the latter in a `models` (or similar) module makes it a bit easier to reuse them in other contexts.

trishankatdatadog committed:
update hvac

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
update how salt lengths are computed

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

sechkova commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,289 @@
+# Classes.
+
+
+class Metadata(ABC):
```
I don't think any ABC features are used anymore?

lukpueh replied:
Yes, let's remove.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,289 @@
+    def sign(self, key_ring: KeyRing) -> JsonDict:
+        # FIXME: Needs documentation of expected behavior
+        signed_bytes = self.signed_bytes
+        signatures = self.__signatures
```
Is there any intention behind` self.__signatures` and `self.signatures` both existing?

lukpueh replied:
I don't think so. I haven't touched the sign/verify functions yet, because I wasn't sure about the original intended behavior (that's why I left above comment there).

But I think we agreed that my plans for these functions as outlined in https://github.com/theupdateframework/tuf/pull/1060#issuecomment-660056376 seemed reasonable.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -0,0 +1,289 @@
+            self.__update_signature(signatures, key.keyid, signature)
+
+        self.__signatures = signatures
+        return self.signable
```
`self.signable` seems like a leftover?

sechkova commented on tuf/api/metadata.py:
I would expect `read_from_json` and `write_to_json` to be part of the same `Metadata` class. What do you think?

lukpueh replied:
I agree, the only problem is that `read_from_json`, as opposed to `write_to_json`, requires knowledge about the class it should instantiate an object from for the `signed` field. That's why it made sense to me that this method is on the Signed class (i.e. any of Target/Timestamp/Snapshot).

But it would be easy to also have a `Metadata.read_from_json` that instantiates the corresponding child class of `Signed`, based on the "signed.type" JSON field.

That's what we do in [in-toto](https://github.com/in-toto/in-toto/blob/73dcfc4a3e8ae739873377883c3a5a6cb3e15099/in_toto/models/metadata.py#L116-L120).

lukpueh replied:
FYI https://github.com/theupdateframework/tuf/pull/1112/commits/41fe57f58b9351d1bcfbb20f55be0d1be351c6d0

sechkova commented:
AFAICT initially this API was created having in mind easily updating a single targets metadata but if we consider it as a base of a tuf rework we need an implementation of a `Root(Signed)` class too.

joshuagl commented:
The core feature of this PR, the metadata API, has landed in #1112. That leaves the Vault functionality of this PR to resolve.

I think securesystemslib would be a good home for it, but either way. Should we close this PR and open an Issue in sslib or here to request the Vault feature with a link to this PR for a reference implementation?

trishankatdatadog commented:
> The core feature of this PR, the metadata API, has landed in #1112. That leaves the Vault functionality of this PR to resolve.
> 
> I think securesystemslib would be a good home for it, but either way. Should we close this PR and open an Issue in sslib or here to request the Vault feature with a link to this PR for a reference implementation?

Yes, please, we should upstream the Hashicorp Vault functionality against securesystemslib

joshuagl commented:
Closing this PR as the metadata API landed in #1112. 

Cross-linking to a securesystemslib Issue requesting an enhancement to support Hashicorp Vault: secure-systems-lab/securesystemslib#307
